source/UI/UIUpgradePanel/UIUpgradeShop.gd

```diff
@@ -0,0 +1,62 @@
+class_name UpgradeShop
+extends Control
+
+export var _upgrade_button_scene: PackedScene
+
+onready var _container := $Panel/VBoxContainer
+onready var _anim_player := $AnimationPlayer
+
+var player: Player setget set_player
+
+func set_player(new_player: Player) -> void:
+	player = new_player
+	player.connect("gold_changed", self, "_on_Player_gold_changed")
+
+
+func add_upgrade_option(upgrade: Upgrade) -> void:
+	var button: Button = _upgrade_button_scene.instance()
+	_container.add_child(button)
+
+	button.connect("pressed", self, "_on_Button_pressed", [button])
+	button.upgrade = upgrade
+	button.disabled = player.gold < upgrade.cost
```

Comment:
You can directly bind the `upgrade` to the call, that way, you don't have to store any reference to the upgrade.